node/node.go

```diff
@@ -374,76 +337,71 @@ func (node *Node) Start() error {
 			return err
 		}
 		topicNamed := allTopics[i].Name
+
+		if nodeconfig.GroupID(topicNamed) == nodeconfig.NewGroupIDByShardID(
+			nodeconfig.ShardID(node.Consensus.ShardID),
+		) {
+
+			pubsub.RegisterTopicValidator(
+				topicNamed,
+				func(ctx context.Context, peer libp2p_peer.ID, msg *libp2p_pubsub.Message) bool {
+					sig2 := msg.GetData()
+					if len(sig2) < p2pMsgPrefixSize {
+						return true
+					}
+
+					cnsMsg := sig2[p2pMsgPrefixSize:][proto.MessageCategoryBytes:]
+
+					var (
+						m         msg_pb.Message
+						senderKey *bls.PublicKey
+					)
+
+					if err := protobuf.Unmarshal(cnsMsg, &m); err != nil {
+						errChan <- err
+						return false
+					}
+
+					if c := m.GetConsensus(); c.GetSenderPubkey() != nil {
+						key, err := bls_cosi.BytesToBLSPublicKey(
+							c.GetSenderPubkey(),
+						)
+						if err != nil {
+							errChan <- err
+							return false
+						}
+						senderKey = key
+					}
+
+					if err := consensus.VerifyMessageSig(senderKey, &m); err != nil {
```

Comment:
any tolerance level on the sender? what if the peer had some outdated db? will you blacklist the peer forever? This has to be very careful.

Reply:
It’s a time tolerance already, time expires, you have to dig into source of it to see. 

Db would be outdated by over an epoch for this to happen, can you say a specific scenario in which it can happen